### PR TITLE
Allow NatSpec comments for state variables

### LINKED ISSUE #3418: Allow NatSpec comments for variables
# Test case

```solidity
pragma solidity ^0.5.0;

/// @title A simulator for Bug Bunny, the most famous Rabbit
/// @author Warned Bros
/// @notice You can use this contract for only the most basic simulation
/// @dev All function calls are currently implement without side effects
contract BugsBunny {
    /// @author Bob Clampett
    /// @notice Determine if Bugs will or will not accept given food
    /// @dev Food must be provided in UTF-8 format
    /// @param _food The name of a food to evaluate (English)
    /// @return true if Bugs will eat it, false otherwise
    function doesEat(string calldata _food) external pure returns (bool) {
        return keccak256(bytes(_food)) == keccak256("carrot");
    }
    
    /// @title example of title
    /// @author example of author
    /// @notice example of notice
    /// @dev example of dev
    /// @param example of param
    /// @return example of return
    string public constant name = "Bugs";
}
```

Run the command:

```sh
solc -o outputDirectory --userdoc --devdoc tmp.sol
```

Then test output with:

```sh
jq .methods outputDirectory/*
```

# Expected outcome

Documentation is output for `doesEat` and `name` public functions.

# Actual outcome

```jsons
{
  "doesEat(string)": {
    "author": "Bob Clampett",
    "details": "Food must be provided in UTF-8 format",
    "params": {
      "_food": "The name of a food to evaluate (English)"
    },
    "return": "true if Bugs will eat it, false otherwise"
  }
}
{
  "doesEat(string)": {
    "notice": "Determine if Bugs will or will not accept given food"
  }
}
```

# Follow on work

- [ ] Update documentation at https://github.com/ethereum/wiki/wiki/Ethereum-Natural-Specification-Format

### PR DESCRIPTION
**Pull request by erak**
Fixes https://github.com/ethereum/solidity/issues/3418.

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 1]
```diff
@@ -307,10 +307,11 @@ bool ReferencesResolver::visit(Return const& _return)
 
 void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 {
+	bool isStateVariable = dynamic_cast<VariableDeclaration const*>(&_variable);
```
Comment (ekpyron) [inline libsolidity/analysis/ReferencesResolver.cpp]:
```suggestion
	bool isStateVariable = dynamic_cast<StateVariableDeclaration const*>(&_variable);
```
?

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/ReferencesResolver.cpp (hunk 1 above)
Comment (erak) [reply libsolidity/analysis/ReferencesResolver.cpp]: That's a defect ;-) Thanks!

Comment (ekpyron) [reply libsolidity/analysis/ReferencesResolver.cpp]: It's interesting compiler errors in CI: ``the compiler can assume that the address of ‘_variable’ will never be NULL`` - never seen those, but it's legit :-)!

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `ea0d33199489956e2f7c97a1861c9b673fd98c04`:
```
Coding style error:
 libsolidity/ast/AST.h:825: ASTPointer<StructuredDocumentation> const& _documentation,
 libsolidity/ast/AST.h:833: StructurallyDocumented(_documentation),
 libsolidity/ast/ASTJsonImporter.cpp:420: _node["documentation"].isNull() ? nullptr : createDocumentation(member(_node, "documentation")),
 libsolidity/parsing/Parser.cpp:681: ASTPointer<StructuredDocumentation> documentation = parseStructuredDocumentation();
 libsolidity/parsing/Parser.cpp:795: documentation,
 libsolidity/parsing/Parser.cpp:1554: Visibility::Default,
 libsolidity/parsing/Parser.cpp:1555: nullptr

```
Please check that your changes are working as intended.

**Comment by erak** [comment]
I think that this goes into the right direction now. There are some parts which need some more thoughts (e.g. how to handle comments on non-state-variables, see commented code in `Parser.cpp`), but I'm not able to continue on this any further. Would be good, if someone else could take over.

**Comment by chriseth** [comment]
Replaced by #8532

**Pull request closed without merging**